### PR TITLE
ISPN-8794 Hot Rod client security configuration usability simplification

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-8794

ISPN-8794 Improve Hot Rod client security configuration usability
    
    - don't require calling enable()
    - default realm is ApplicationRealm
    - default mech is DIGEST-MD5
    - single-file trust certificate chain in PEM format

ISPN-8794 Restructure Hot Rod security documentation
    
    - Move client security from server guide to user guide
    - Adapt examples to simplified configuration API

### CODE REVIEW

File: documentation/src/main/asciidoc/user_guide/server_protocols_hotrod.adoc [hunk 1]
```diff
@@ -83,6 +84,284 @@ which means that the value should be taken from a system property named jboss.bi
...
+==== Authentication
+
+If the server has set up authentication, you need to configure your client accordingly. Depending on the mechs enabled on the server,
+the client must provide the required information.
+
+NOTE: This section is about client configuration. If you want to set up the server to require authentication, read the
+link:link:../server_guide/server_guide.html#security:hotrod_auth[Hot Rod server authentication] section.
+
+===== DIGEST-MD5
+DIGEST-MD5 is the recommended approach for simple username/password authentication scenarios. If you are using the default
```
Comment (gustavonalle) [inline documentation/src/main/asciidoc/user_guide/server_protocols_hotrod.adoc]: IMO we should document that MD5 should not be used, and also we should not use it as a default algorithm.

**Comment by tristantarrant** [comment]
Digest-md5 is not particularly insecure. It is the default mech used by
LDAP servers. It is a multiple stage mech with a nonce that protects
against Replay attacks. Only enabling TLS provides additional security.

**Comment by gustavonalle** [comment]
@tristantarrant an MD5 digest can be reverted, so I suppose the issue is not only replay attacks. Please refer to https://en.wikipedia.org/wiki/MD5

"It can be cracked by brute-force attack and suffers from extensive vulnerabilities as detailed in the security section below"

**Comment by gustavonalle** [comment]
@tristantarrant Maybe the use of MD5 in the Digest-md5 is not as bad as using "pure" MD5, but since MD5 is broken, isn't it better to pick a stronger default digest?

**Comment by gustavonalle** [comment]
@tristantarrant One more reference (July 2011):

https://tools.ietf.org/html/rfc6331

"This memo describes problems with the DIGEST-MD5 Simple
   Authentication and Security Layer (SASL) mechanism as specified in
   RFC 2831.  It marks DIGEST-MD5 as OBSOLETE in the IANA Registry of
   SASL mechanisms and moves RFC 2831 to Historic status."

**Comment by ryanemerson** [comment]
@tristantarrant The build is failing due to a checkstyle error:

```
[ERROR] /home/infinispan/workspace/Infinispan_PR-5748-I2535JS43ZAP7JVIOGBWCZ4D7C5CDSYI7TGQJQ24AHD774L5DEWA/client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/TypedProperties.java:3:8: Unused import - java.util.Optional. [UnusedImports]
```

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelInitializer.java [hunk 2]
```diff
@@ -117,6 +118,9 @@ private void initSsl(Channel channel) {
                      ssl.trustStorePassword(),
                      configuration.classLoader()));
             }
+            if (ssl.trustStorePath() != null) {
+               builder.trustManager(new File(ssl.trustStorePath()));
```
Comment (ryanemerson) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transport/netty/ChannelInitializer.java]: Much easier than doing it ourselves :+1:

**Comment by tristantarrant** [comment]
We can definitely use SCRAM, but we probably need to add wildfly-security
to the client dependencies as well as ensuring that the c++ cliente has an
implementation of it.

**Comment by tristantarrant** [comment]
Created https://issues.jboss.org/browse/ISPN-8810 to track moving to SCRAM and fixed the unused import

**Comment by ryanemerson** [comment]
@tristantarrant Some of the test failures are related:

```java
org.infinispan.jcache.remote.JCacheConfigurationPropertiesFileTest.testPropertiesConfiguration
org.infinispan.spring.provider.SpringRemoteCacheManagerFactoryBeanTest.infinispanRemoteCacheFactoryBeanShouldReportTheMostDerivedObjectType
org.infinispan.spring.provider.SpringRemoteCacheManagerFactoryBeanTest.destroyShouldStopTheProducedCache
org.infinispan.spring.config.InfinispanRemoteCacheManagerDefinitionTest.springTestContextPrepareTestInstance
org.infinispan.spring.provider.SpringRemoteCacheManagerFactoryBeanTest.shouldProduceARemoteCacheManagerConfiguredUsingDefaultSettingsIfNeitherConfigurationPropertiesNorConfigurationPropertiesFileLocationHasBeenSet
org.infinispan.spring.support.remote.InfinispanRemoteCacheManagerFactoryBeanTest.shouldProduceARemoteCacheManagerConfiguredUsingDefaultSettingsIfNeitherConfigurationPropertiesNorConfigurationPropertiesFileLocationHasBeenSet
org.infinispan.spring.support.remote.InfinispanRemoteCacheManagerFactoryBeanContextTest.springTestContextPrepareTestInstance
org.infinispan.spring.provider.SpringRemoteCacheManagerFactoryBeanTest.shouldProduceAStoppedCacheIfStartAutomaticallyIsSetToFalse
org.infinispan.spring.support.remote.InfinispanRemoteCacheManagerFactoryBeanTest.shouldProduceAStoppedCacheIfStartAutomaticallyIsSetToFalse
org.infinispan.spring.provider.SpringRemoteCacheManagerFactoryBeanContextTest.springTestContextPrepareTestInstance
org.infinispan.spring.support.remote.InfinispanRemoteCacheManagerFactoryBeanTest.destroyShouldStopTheProducedCache
org.infinispan.spring.support.remote.InfinispanRemoteCacheManagerFactoryBeanTest.infinispanRemoteCacheFactoryBeanShouldReportTheMostDerivedObjectType
```

**Comment by tristantarrant** [comment]
@ryanemerson failures are unrelated

**Comment by galderz** [comment]
Needs rebase

**Comment by tristantarrant** [comment]
Rebased. It was just a conflict in the docs, so no need to wait for CI again

**Pull request merged**